csrc/multidevice/utils.cpp

```diff
@@ -96,4 +99,58 @@ bool isResharding(Expr* expr) {
   return !haveDifferentSharding(tv_ref, tvs).empty();
 }
 
+namespace {
+
+void shardAllLike(TensorView* ref, std::vector<TensorView*> tvs) {
+  for (auto tv : tvs) {
+    tv->setDeviceMesh(ref->getDeviceMesh());
+  }
+  scheduler_utils::parallelizeAllLike(ref, tvs, {ParallelType::DIDx});
+}
+
+void reshardBefore(Expr* expr, Fusion* fusion) {
+  NVF_ERROR(
+      expr->outputs().size() == 1,
+      "multi-output expressions are not supported");
+  NVF_ERROR(
+      expr->outputs().at(0)->isA<TensorView>(),
+      "the expression's output is not a TensorView");
+  TensorView* output = expr->outputs().at(0)->as<TensorView>();
+  std::unordered_set<TensorView*> inputs;
+  std::transform(
+      expr->inputs().begin(),
+      expr->inputs().end(),
+      std::inserter(inputs, inputs.end()),
+      [](Val* val) {
+        NVF_ERROR(
+            val->isA<TensorView>(),
+            "the expression's input is not a TensorView");
+        return val->as<TensorView>();
+      });
+  std::vector<TensorView*> new_inputs;
+  for (auto input : haveDifferentSharding(output, inputs)) {
+    // TODO: reuse cacheAfter?
+    // TODO: here we should add a mechanism to potentially reuse the inserted
+    // resharding accross all the consumer of the resharded tensor. This way we
+    // could avoid wasteful resharding set insertion.
+    TensorView* new_input = set(input);
+    expr = ir_utils::replaceValInExprInputs(expr, input, new_input);
+    new_inputs.push_back(new_input);
+  }
+  if (!new_inputs.empty()) {
+    shardAllLike(output, new_inputs);
+  }
+}
+
+} // namespace
+
+void insertReshardings(Fusion* fusion) {
+  auto exprs = fusion->exprs();
+  for (auto expr : exprs) {
+    if (!isLowerableToCommunication(expr)) {
+      reshardBefore(expr, fusion);
```

Comment:
There is a small bug here - isLowerableToCommunication marks all reductions as lowerable, but that is only true if the reduction is over a device axis otherwise it needs a reshard expression.

Reply:
fixed